src/NServiceBus.Core/Persistence/InMemory/SagaPersister/InMemorySagaPersister.cs

```diff
@@ -131,33 +115,43 @@ where s.Value.SagaData.GetType() == sagaType && (s.Key != saga.Id)
 
         ConcurrentDictionary<Guid, VersionedSagaEntity> data = new ConcurrentDictionary<Guid, VersionedSagaEntity>();
 
-        int version;
-
         class VersionedSagaEntity
         {
             public VersionedSagaEntity(IContainSagaData sagaData, VersionedSagaEntity original = null)
             {
                 SagaData = DeepClone(sagaData);
-                versionCache = original?.versionCache ?? new ConditionalWeakTable<IContainSagaData, SagaVersion>();
+                if (original != null)
+                {
+                    original.ConcurrencyCheck(sagaData);
+
+                    versionCache = original.versionCache;
+                    version = original.version;
+                    Interlocked.Increment(ref version);
```

Comment:
I agree. We do not need `Interlocked`. All invocations of constructor or synchronized by `ConcurrentDictionary`